README.md
# Solr Scripts

## About

I do quite a bit of work with SOLR (Cloud).  There are a few actions that happen over and over.  I created these scripts to handle the repetitiveness.

### Be sure to see the [Changelog](https://github.com/likethecolor/solr-scripts/blob/master/CHANGELOG.md)

## Scripts

- sc-create-alias.sh

 Script for creating an alias.
 https://cwiki.apache.org/confluence/display/solr/Collections+API#CollectionsAPI-api4

    - -a alias-name: name of alias [default:none]
    - -c collection: name of collection [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-create-alias...out]

- sc-delete-alias.sh

 Script for deleting an alias.
 https://cwiki.apache.org/confluence/display/solr/Collections+API#CollectionsAPI-api5

    - -a alias-name: name of alias [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]

- sc-create-collection.sh

 Script for creating a given collection.
 https://cwiki.apache.org/confluence/display/solr/Collections+API#CollectionsAPI-api1

    - -c collection: name of collection [default:none] 
    - -g config: name of the zookeeper configuraiton as found in zookeeper:/configs [default:none] 
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-create-collection...out]

- sc-delete-collection.sh

 Script for deleting a given collection.
 https://cwiki.apache.org/confluence/display/solr/Collections+API#CollectionsAPI-api6

    - -c collection: name of collection [default:none]
    - -d: give to delete the index files [default:false]
    - -h host:port: host and port where zookeeper is running [default:none]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-delete-collection...out]

- sc-delete-data.sh

 Script for deleting all data of a given collection.

    - -c collection: name of collection [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-delete-data....]

- sc-update-data.sh

 Script for updating a value in a document.

    - -c collection: name of collection [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]
    - -f field-name: name of the field whose value is to be modified [default:none]
    - -i unique id field: name of the unique id field of the document to change [default:none]
    - -n new value: value to change to [default:none]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-update-data....out]
    - -u unique id value: value of the unique id of the document to change [default:none]

- sc-delete-replica.sh

 Script for deleting a replica from a given shard from a give collection.
 https://cwiki.apache.org/confluence/display/solr/Collections+API#CollectionsAPI-api9

    - -c collection: name of collection [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:8983]
    - -o file: path to file to which output will be written [default:mktemp -q /tmp/sc-delete-data....]
    - -r replica: name of replica to delete
    - -s shard: name of shard to which the replica belongs


- zk-upload-config.sh

 Script for uploading solr conf directory to zookeeper

    - -d path: path to solr's conf directory [default:none]
    - -h host:port: host and port where zookeeper is running [default:localhost:2181]
    - -l path: path to solr's lib where the following jars are located [default:none]

  |    jars                 |
  |:-----------------------:|
  | commons-cli-1.2.jar     |
  | commons-io-2.1.jar      |
  | log4j-1.2.16.jar        |
  | slf4j-api-1.6.6.jar     |
  | slf4j-log4j12-1.6.6.jar |
  | solr-core-4.6.0.jar     |
  | solr-solrj-4.6.0.jar    |
  | zookeeper-3.4.5.jar     |

    - -n name: name of this configuration as it will appear in zookeeper:/configs [default:none]


CHANGELOG.md
# Changelog

## 1.0.2 - Nov 23, 2015
- variable name used was $host, should be $host_port
- added sc-delete-replica.sh

## 1.0.1 - Nov 23, 2015
- the script was lookng for specific versions on the jar files in solr lib
- removed that requirement so that any version of the specific jar names will be used

## 1.0.0 - Nov 19, 2015
- Initial check in of:
    - sc-create-alias.sh
    - sc-create-collection.sh
    - sc-delete-alias.sh
    - sc-delete-collection.sh
    - sc-delete-data.sh
    - sc-update-data.sh
    - zk-upload-config.sh


sc-delete-data.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for deleting all data of a given collection.
#
# Arguments - all are required:
#
#   -c collection: name of collection
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:$SOLR_HOST_PORT]
#
#   -o file: path to file to which output will be written
#                 [default:mktemp -q /tmp/sc-delete-data....]
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0 '.sh'`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi
rm $log_file
log_file=$log_file.log
touch $log_file

function usage() {
  echo "usage: $0 -h host:port -c collection-name -o output-file [--dry-run]"
  echo '  Sends the DELETE action to the host deleting all data in the specified'
  echo '  collection-name.  If --dry-run is provided only the url will be printed,'
  echo '  nothing will be deleted.'
  echo
  echo '  note: the env variable $SOLR_HOST_PORT will be used as default value for -h'
  if test "$@" != ''; then
    echo -e "---> $color_prefix$@$color_suffix"
  fi
}


collection=''
dry_run=0
host_port=$SOLR_HOST_PORT

while test -n "$1"; do
  case "$1" in
    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-c')
      shift
      collection=$1
      shift
      ;;

    '-o')
      shift
      output_file=$1
      shift
      ;;

    '--dry-run')
      shift
      dry_run=1
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

if test -z "$collection"; then
  usage 'no collection name (-c)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$host_port"; then
  if test -z "$SOLR_HOST_PORT"; then
    usage 'no host:port (-h) and no $SOLR_HOST_PORT'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    host_port=$SOLR_HOST_PORT
    log_stdout "using \$SOLR_HOST_PORT env variable: $host_port"
  fi
fi
if test -z "$output_file"; then
  log_info 'no output file (-o)'
  log_info 'using output file based on log file name'
  output_file='/tmp/'$(basename $log_file '.log').out
fi

exec > >(tee $log_file) 2>&1
log_stdout_sep
log_info 'starting delete using:'
log_info "    host:port: $host_port"
log_info "   collection: $collection"
log_info "     log file: $log_file"
log_info "  output file: $output_file"
log_stdout_sep

delete_url="http://${host_port}/solr/${collection}/update/?stream.body=<delete><query>*:*</query></delete>&commit=true&optimize=true"

exit_code=1
client=$(which curl)
if test -n "$client"; then
  log_info "found curl: $client"
  if test $dry_run -eq 0; then
    curl $delete_url -s --output $output_file 2>&1
    exit_code=$?
    if test $exit_code -eq 0; then
      if grep -q ERROR $output_file; then
        log_error "there are errors in the output file: $output_file"
        exit_code=1
      fi
    fi
  else
    log_info 'DRY RUN'
    log_info $delete_url
    exit_code=0
  fi
else
  log_error 'cannot find curl in $PATH'
fi

if test $exit_code -eq 0; then
  log_info 'success'
else
  log_error 'failed'
fi
log_info 'done'
exit $exit_code


sc-update-data.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for updating a value in a document.
#
# Arguments - all are required:
#
#   -c collection: name of collection
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:$SOLR_HOST_PORT]
#
#   -f field-name: name of the field whose value is to be modified
#                 [default:none]
#
#   -i unique id field: name of the unique id field of the document to change
#                 [default:none]
#
#   -n new value: value to change to
#                 [default:none]
#
#   -o file: path to file to which output will be written
#                 [default:mktemp -q /tmp/sc-update-data....out]
#
#   -u unique id value: value of the unique id of the document to change
#                 [default:none]
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0 '.sh'`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi
rm $log_file
log_file=$log_file.log
touch $log_file

function usage() {
  echo "usage: $0 -c collection-name -h host:port -f field-name -i id-field -n new-value -o output-file -u unique-id"
  echo "  Change the value of a solr document.  Given the unique id (-u) and that id's"
  echo '  field name (-i) change the value of the field (-f) to the new value (-n).  The'
  echo '  output of the call will be stored in the output file (-o).'
  echo
  echo '  note: the env variable $SOLR_HOST_PORT will be used as default value for -h'
  if test "$@" != ''; then
    echo -e "---> ${color_prefix}$@${color_suffix}"
  fi
}

collection=my-collection
host_port=$SOLR_HOST_PORT
field_name=''
new_value=''
unique_id_field_name=''
unique_id_field_value=''

while test -n "$1"; do
  case "$1" in
    '-c')
      shift
      collection=$1
      shift
      ;;

    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-f')
      shift
      field_name=$1
      shift
      ;;

    '-i')
      shift
      unique_id_field_name=$1
      shift
      ;;

    '-n')
      shift
      new_value=$1
      shift
      ;;

    '-o')
      shift
      output_file=$1
      shift
      ;;

    '-u')
      shift
      unique_id_value=$1
      shift
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

# verify that all args are not empty
#
if test -z "$collection"; then
  usage 'no collection (-c)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$field_name"; then
  usage 'no field name (-f)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$host_port"; then
  if test -z "$SOLR_HOST_PORT"; then
    usage 'no host:port (-h) and no $SOLR_HOST_PORT'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    host_port=$SOLR_HOST_PORT
    log_stdout "using \$SOLR_HOST_PORT env variable: $host_port"
  fi
fi
if test -z "$output_file"; then
  log_info 'no output file (-o)'
  log_info 'using output file based on log file name'
  output_file='/tmp/'$(basename $log_file '.log').out
fi
if test -z "$new_value"; then
  usage 'no new value (-n)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$unique_id_field_name"; then
  usage 'no unique id field name (-i)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$unique_id_value"; then
  usage 'no unique field value (-u)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi

exec > >(tee $log_file) 2>&1
log_stdout_sep
log_stdout 'starting to change value using:'
log_stdout "             collection: $collection"
log_stdout "              host:port: $host_port"
log_stdout "             field name: $field_name"
log_stdout "              new value: $new_value"
log_stdout "               log file: $log_file"
log_stdout "            output file: $output_file"
log_stdout "   unique id field name: $unique_id_field_name"
log_stdout "  unique id field value: $unique_id_field_value"
log_stdout_sep

exit_code=1
client=$(which curl)
if test -n "$client"; then
  log_info "found curl: $client"
  log_info curl http://$host_port/solr/$collection/update?commit=true -s --output $output_file -H 'Content-Type: text/xml' --data-binary "<add><doc><field name=\"$unique_id_field_name\">$unique_id_value</field><field name=\"$field_name\" update=\"set\">$new_value</field></doc></add>"
  curl http://$host_port/solr/$collection/update?commit=true -s --output $output_file -H 'Content-Type: text/xml' --data-binary "<add><doc><field name=\"$unique_id_field_name\">$unique_id_value</field><field name=\"$field_name\" update=\"set\">$new_value</field></doc></add>" 2>&1
  exit_code=$?
  log_info "curl call returned: $exit_code"
  if test $exit_code -eq 0; then
    if grep -q ERROR $output_file; then
      log_error "there are errors in the output file: $output_file"
      exit_code=1
    fi
  fi
else
  log_error 'cannot find curl in $PATH'
  exit_code=1
fi

if test $exit_code -eq 0; then
  log_info 'success'
else
  log_error 'failed'
fi
log_stdout 'done'
exit $exit_code


sc-delete-alias.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for deleting an alias.
#
# Arguments - all are required:
#
#   -a alias-name: name of alias
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:$SOLR_HOST_PORT]
#
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0 '.sh'`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi
rm $log_file
log_file=$log_file.log
touch $log_file

function usage() {
  echo "usage: $0 -a alias-name -h host:port"
  echo '  Sends the DELETEALIAS action to the host deleting an existing alias having the'
  echo '  name alias-name.'
  echo
  echo '  note: the env variable $SOLR_HOST_PORT will be used as default value for -h'
  if test "$@" != ''; then
    echo -e "---> ${color_prefix}$@$color_suffix"
  fi
}

alias=''
host_port=$SOLR_HOST_PORT

while test -n "$1"; do
  case "$1" in
    '-a')
      shift
      alias=$1
      shift
      ;;

    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-o')
      shift
      output_file=$1
      shift
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

if test -z "$alias"; then
  usage 'no alias-name (-a)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$host_port"; then
  if test -z "$SOLR_HOST_PORT"; then
    usage 'no host:port (-h) and no $SOLR_HOST_PORT'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    host_port=$SOLR_HOST_PORT
    log_stdout "using \$SOLR_HOST_PORT env variable: $host_port"
  fi
fi
if test -z "$output_file"; then
  log_info 'no output file (-o)'
  log_info 'using output file based on log file name'
  output_file='/tmp/'$(basename $log_file '.log').out
fi

exec > >(tee $log_file) 2>&1
log_info 'starting to delete existing alias using:'
log_info "    host:port: $host_port"
log_info "   alias name: $alias"
log_info "     log file: $log_file"
log_info "  output file: $output_file"


url="${host_port}/solr/admin/collections?action=DELETEALIAS&name=${alias}"
exit_code=1
client=$(which curl)
if test -n "$client"; then
  log_info "found curl: $client"
  log_info "curl '$url' --output '$output_file'"
  curl "$url" -s --output "$output_file" 2>&1
  exit_code=$?
  if test $exit_code -eq 0; then
    if grep -q ERROR $output_file; then
      log_error "there are errors in the output file: $output_file"
      exit_code=1
    fi
  fi
fi

if test $exit_code -eq 0; then
log_info 'success'
else
  log_error 'failed'
fi
log_info 'done'
exit $exit_code


zk-upload-config.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for uploading solr conf directory to zookeeper
#
# Arguments - all are required:
#
#   -d path: path to solr's conf directory
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:$ZK_HOST_PORT]
#
#   -l path: path to solr's lib where the following jars are located
#              - commons-cli-*.jar
#              - commons-io-*.jar
#              - logj-*.jar
#              - slfj-api-*.jar
#              - slfj-logj-*.jar
#              - solr-core-*.jar
#              - solr-solrj-*.jar
#              - zookeeper-*.jar
#                 [default:$SOLR_ZK_LIB_DIR]
#
#   -n name: name of this configuration as it will appear in zookeeper:/configs
#                 [default:none]
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi

function usage() {
  echo "usage: $0 -d conf-dir -h host:port -l solr-lib -n conf-name"
  echo '  Uploads to zookeeper running on host:port the configuration directory located'
  echo '  at conf-dir giving the configuration the name conf-name.  The jars found in -l'
  echo '  are used to do the uploading.'
  if test "$@" != ''; then
    echo -e "---> ${color_prefix}$@$color_prefix"
  fi
}


# jars used to perform the upload
#
jars="commons-cli- \ 
commons-io- \
log4j- \
slf4j-api \
slf4j-log4j12- \
solr-core- \
solr-solrj- \
zookeeper-"

# defaults
#
conf_dir=''
conf_name=''
host_port=$ZK_HOST_PORT
solr_lib=$SOLR_ZK_LIB_DIR

while test -n "$1"; do
  case "$1" in
    '-d')
      shift
      conf_dir=$1
      shift
      ;;

    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-l')
      shift
      solr_lib=$1
      shift
      ;;

    '-n')
      shift
      conf_name=$1
      shift
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

# verify that all args are not empty
#
if test -z "$conf_dir"; then
  usage 'no conf-dir (-d)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$conf_name"; then
  usage 'no conf-name (-n)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$host_port"; then
  if test -z "$ZK_HOST_PORT"; then
    usage 'no host:port (-h) and no $ZK_HOST_PORT'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    host_port=$ZK_HOST_PORT
    log_stdout "using \$ZK_HOST_PORT env variable: $host_port"
  fi
fi
if test -z "$solr_lib"; then
  if test -z "$SOLR_ZK_LIB_DIR"; then
    usage 'no solr_lib (-l) and no $SOLR_ZK_LIB_DIR'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    solr_lib=$SOLR_ZK_LIB_DIR
    log_stdout "using \$SOLR_ZK_LIB_DIR env variable: $solr_lib"
  fi
fi

exec > >(tee $log_file) 2>&1
log_stdout 'starting to upload conf'
log_stdout "using host:port: $host_port"
log_stdout " using conf-dir: $conf_dir"
log_stdout "using conf-name: $conf_name"
log_stdout "using  solr-lib: $solr_lib"

# create classpath
cp=''
for jar in $solr_lib/*;
do
  for patt in $jars
  do
    if [[ $jar =~ $patt ]]; then
      if test -n "$cp"; then
        cp="$cp:"
      fi
      cp="$cp$jar"
    fi
  done
done

# RUN!
#
cmd="java -cp $cp org.apache.solr.cloud.ZkCLI -cmd upconfig -zkhost ${host_port} -confdir ${conf_dir} -confname ${conf_name}"
log_stdout $cmd
$cmd
exit_code=$?

if test $exit_code -eq 0; then
  log_info 'success'
else
  log_error 'failed'
fi
log_stdout 'done'
exit $exit_code


sc-delete-replica.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for deleting a given replica
#
# Arguments - all are required:
#
#   -c collection: name of collection containing replica
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:$SOLR_HOST_PORT]
#
#   -o file: path to file to which output will be written
#                 [default:mktemp -q /tmp/sc-delete-collection...out]
#
#   -r replica: name of replica
#                 [default:none]
#
#   -s shard: shard number
#                 [default:none]
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0 '.sh'`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi
rm $log_file
log_file=$log_file.log
touch $log_file

function usage() {
  echo "usage: $0 -c collection -h host:port -o output-file -r replica -s shard"
  echo '  sends the DELETEREPLICA action to the host deleting an existing'
  echo '  replica within the given collection having the name replica.'
  echo
  echo '  note: the env variable $SOLR_HOST_PORT will be used as default value for -h'
  if test "$@" != ''; then
    echo -e "---> ${color_prefix}$@$color_suffix"
  fi
}

collection=''
host_port=$SOLR_HOST_PORT
output_file=''
replica=''
shard=''

while test -n "$1"; do
  case "$1" in
    '-c')
      shift
      collection=$1
      shift
      ;;

    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-o')
      shift
      output_file=$1
      shift
      ;;

    '-r')
      shift
      replica=$1
      shift
      ;;

    '-s')
      shift
      shard=$1
      shift
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

if test -z "$host_port"; then
  if test -z "$SOLR_HOST_PORT"; then
    usage 'no host:port (-h) and no $SOLR_HOST_PORT'
    echo -e "${color_prefix}exiting$color_suffix"
  else
    host_port=$SOLR_HOST_PORT
    log_stdout "using \$SOLR_HOST_PORT env variable: $host_port"
  fi
fi
if test -z "$collection"; then
  usage 'no collection (-c)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$output_file"; then
  log_info 'no output file (-o)'
  log_info 'using output file based on log file name'
  output_file='/tmp/'$(basename $log_file '.log').out
fi
if test -z "$replica"; then
  usage 'no replica (-r)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$shard"; then
  usage 'no shard (-s)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi

exec > >(tee $log_file) 2>&1
log_stdout 'starting to delete existing core using:'
log_stdout "     host:port: $host_port"
log_stdout "     core name: $core"
log_stdout "      log file: $log_file"
log_stdout "   output file: $output_file"
log_stdout "       replica: $replica"
log_stdout "         shard: $shard"
url="${host_port}/solr/admin/collections?action=DELETEREPLICA&collection=${collection}&shard=${shard}&replica=${replica}"

exit_code=1
client=$(which curl)
if test -n "$client"; then
  log_info "found curl: $client"
  log_stdout "curl '$url'"
  curl "$url" -s --output $output_file
  exit_code=$?
  log_info "curl call returned: $exit_code"
  if test $exit_code -eq 0; then
    if grep -q ERROR $output_file; then
      log_error "there are errors in the output file: $output_file"
      exit_code=1
    fi
  fi
else
  log_error 'cannot find curl in $PATH'
  exit_code=1
fi

if test $exit_code -eq 0; then
  log_info 'success'
else
  log_error 'failed'
fi
log_stdout 'done'
exit $exit_code


sc-create-alias.sh
#!/bin/bash

# Copyright (c) 2015.  Dan Brown <dan@likethecolor.com>
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy
# of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.
#
# Original author: Dan Brown <dan@likethecolor.org>
#
# Script for creating an alias.
#
# Arguments - all are required:
#
#   -a alias-name: name of alias
#                 [default:none]
#
#   -c collection: name of collection
#                 [default:none]
#
#   -h host:port: host and port where zookeeper is running
#                 [default:localhost:8983]
#
#   -o file: path to file to which output will be written
#                 [default:mktemp -q /tmp/sc-create-alias...out]
#

dir=$(cd $(dirname $0); pwd)
source $dir/log

color_prefix='\033[40;31m'
color_suffix='\033[0m'

# create temporary file that will be used to log the actions of this script
#
log_file_base=`basename $0 '.sh'`
log_file=`mktemp -q /tmp/${log_file_base}.XXXXXX`
if test $? -ne 0; then
  echo -e "$0: ${color_prefix}Can't create temp file, exiting...$color_suffix"
  exit 1
fi
rm $log_file
log_file=$log_file.log
touch $log_file

alias=''
collection=''
host_port=''
output_file=''

function usage() {
  echo "usage: $0 -a alias-name -c collection -h host:port"
  echo '  Sends the CREATEALIAS action to the host creating an alias for collection'
  echo '  with the alias name alias-name.'
  if test "$@" != ''; then
    echo -e "---> ${color_prefix}$@${color_suffix}"
  fi
}

while test -n "$1"; do
  case "$1" in
    '-a')
      shift
      alias=$1
      shift
      ;;

    '-c')
      shift
      collection=$1
      shift
      ;;

    '-h')
      shift
      host_port=$1
      shift
      ;;

    '-o')
      shift
      output_file=$1
      shift
      ;;

    *)
      usage "cannot understand argument $1"
      echo -e "${color_prefix}exiting$color_suffix"
      exit 1
      ;;
  esac
done

if test -z "$alias"; then
  usage 'no alias (-a)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$collection"; then
  usage 'no collection (-c)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$host_port"; then
  usage 'no host:port (-h)'
  echo -e "${color_prefix}exiting$color_suffix"
  exit 1
fi
if test -z "$output_file"; then
  log_info 'no output file (-o)'
  log_info 'using output file based on log file name'
  output_file='/tmp/'$(basename $log_file '.log').out
fi

exec > >(tee $log_file) 2>&1
log_stdout 'starting to create alias using:'
log_stdout "    host:port: $host_port"
log_stdout "   collection: $collection"
log_stdout "   alias name: $alias"
log_stdout "     log file: $log_file"
log_stdout "  output file: $output_file"

url="${host_port}/solr/admin/collections?action=CREATEALIAS&name=${alias}&collections=${collection}"

exit_code=1
client=$(which curl)
if test -n "$client"; then
  log_info "found curl: $client"
  log_stdout "curl '$url' -s --output '$output_file'"
  curl "$url" -s --output "$output_file" 2>&1
  exit_code=$?
  log_info "curl call returned: $exit_code"
  if test $exit_code -eq 0; then
    if grep -q ERROR $output_file; then
      log_error "there are errors in the output file: $output_file"
      exit_code=1
    fi
  fi
else
  log_error 'cannot find curl in $PATH'
  exit_code=1
fi

if test $exit_code -eq 0; then
  log_info 'success'
else
  log_error 'failed'
fi
log_stdout 'done'
exit $exit_code
